util.go

```diff
@@ -152,6 +153,22 @@ func (s *segmentedBuffer) Append(input io.Reader, length uint32) error {
 	if n > 0 {
 		s.len += uint32(n)
 		s.cap -= uint32(n)
+		// we are out of cap
+		if len(s.b) == cap(s.b) && s.bPos > 0 {
+			if s.bPos == len(s.b) {
+				// have no unread chunks, just move pos
+				s.bPos = 0
+				s.b = s.b[:0]
+			} else {
```

Comment:
I'd change this to to check if at least half of the slice is free. That'll slightly increase allocations until we hit a steady-state, but should avoid the degenerate case where we slide by one every single time.

Reply:
This is definitely a good improvement, but I am not sure about "at least half of the slice", I think it should be a bit lower than that, how about 0.25 of the capacity? Also, then it will be equal to `append()` growth factor (when cap > 1024). Added with 0.25 for now.

That got me thinking, should we limit the maximum capacity of the buffer (recreate slice with default capacity when reach certain maximum and buffer is empty now)? What is the average Stream lifespan?

I tried to test this on the package's benchmark, but it does not grow at all because of in-memory network.